backend/models/loginRecord.js

```diff
@@ -0,0 +1,127 @@
+ *	GNU Affero General Public License for more details.
+ *
+ *	You should have received a copy of the GNU Affero General Public License
+ *	along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+ "use strict";
+ const db = require("../handler/db");
+ 
+ //detects edge as browser but not device
+ var uaParserJs = require('ua-parser-js');
+ var device = require('device');
+ var ip2location = require('ip-to-location');
+
+ const LoginRecord = {};
+ 
+ LoginRecord.saveLoginRecord = async (req) => {
+        var loginRecord = {
+                _id : req.sessionID,
+                loginTime : new Date(),
+                ipAddr: req.ips[0] || req.ip                    
+        }
+
+        const referrer = req.header('Referer');
+        if(referrer != null){
+                loginRecord.referrer = referrer;
+        }
+
+        const location = await getLocationFromIPAddress(loginRecord.ipAddr);   
+        loginRecord.location = {
+                country: location.country_name,
+                city: location.city
+        };    
+       
+         const userAgentString = req.headers['user-agent'];
+         var uaInfo = {};
+
+         if(isUserAgentFromPlugin(userAgentString)){
+                uaInfo = getUserAgentInfoFromPlugin(userAgentString);
+         }
+         else{
+                uaInfo = getUserAgentInfoFromBrowser(userAgentString);
+         }
+
+         loginRecord.application = uaInfo.application;
+         loginRecord.engine = uaInfo.engine;
+         loginRecord.os = uaInfo.os;
+         loginRecord.device = uaInfo.device;          
+
+         const col = await db.getCollection("loginRecords", req.body.username);
+         await col.insertOne(loginRecord);
+ }
+ 
+ //Format: 
+ //PLUGIN: {OS Name}/{OS Version} {Host Software Name}/{Host Software Version} {Plugin Type}/{Plugin Version}
+ //Example:
+ //PLUGIN: Windows/10.0.19042.0 REVIT/2021.1 PUBLISH/4.15.0
+ function getUserAgentInfoFromPlugin(userAgentString){
+        const userAgentInfo = {};        
+        const userAgentComponents = userAgentString.replace("PLUGIN: ", "").split(' ');                  
+
+        userAgentInfo.application = {
+                name: userAgentComponents[1].split('/')[0],
+                version: userAgentComponents[1].split('/')[1],
+                type: "plugin"        
+        }
+        
+        userAgentInfo.engine = {
+                name: "3drepoplugin",
+                version: userAgentComponents[2].split('/')[1]
+        }
+
+        userAgentInfo.os = {
+                osName: userAgentComponents[0].split('/')[0],
+                osVersion: userAgentComponents[0].split('/')[1]
+        }
+
+        userAgentInfo.device = "desktop";
+
+        return userAgentInfo;
+ }
+
+ function getUserAgentInfoFromBrowser(userAgentString){
+        const userAgentInfo = {};        
+        const userAgentObject = uaParserJs(userAgentString);       
+
+        userAgentInfo.application = {
```

Comment:
you could use the spread operator to copy the object from browser:

```suggestion
        userAgentInfo.application = {
            ...userAgentObject.browser,
            type: "browser"
        }
            
```

Reply:
browser has both major and full version so in that case we might not want to do that ?

![image](https://user-images.githubusercontent.com/71635617/122095144-e7423900-ce04-11eb-87f4-d08677dabd42.png)

Reply:
actually we may as well record both if it gives us both.

the major version is probably more useful for stats purposes but we may want to check specific versions too